### PR TITLE
Adding the windows users virtual table

### PR DESCRIPTION
**Pull request by PoppySeedPlehzr**

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit aba01f81e5044a529e7b4a59dc47f4d1c714a1a5 (Job results: [324](https://jenkins.osquery.io/job/osqueryPullRequestBuild/324/)) failed one or more tests (**Windows**).

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit aba01f81e5044a529e7b4a59dc47f4d1c714a1a5 (Job results: [3320](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3320/)) failed one or more tests (**OS X/Linux**).

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit aba01f81e5044a529e7b4a59dc47f4d1c714a1a5 (Job results: [309](https://jenkins.osquery.io/job/osqueryPullRequestAudit/309/)) failed the code audit.

**Comment by PoppySeedPlehzr** [comment]
```
[+] Running: make format
[+] clang-format (/usr/local/osquery/bin/clang-format) version: clang-format version 3.8.1 (tags/RELEASE_381/final)
clang-format did not modify any files

[+] Running: make check
/bin/bash: cppcheck: command not found
[+] cppcheck () version: 
/bin/bash: cppcheck: command not found
```

File: osquery/tables/system/windows/users.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    if ((pUserIterationBuffer = pUserBuffer) == nullptr ||
+        nEnumStatus != NERR_Success) {
+      break;
+    }
+
+    for (DWORD i = 0; i < dwEntriesRead; i++) {
+      Row r;
+
+      r["username"] = wstring_to_string(pUserIterationBuffer->usri3_name);
+
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: #unneededwhitespace

File: osquery/tables/system/windows/users.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      r["description"] = wstring_to_string(pUserIterationBuffer->usri3_comment);
+      r["uid"] = BIGINT(pUserIterationBuffer->usri3_user_id);
+      r["gid"] = BIGINT(pUserIterationBuffer->usri3_primary_group_id);
+      r["uid_signed"] = BIGINT(pUserIterationBuffer->usri3_user_id);
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Just assign from `r["uid"]`

File: osquery/tables/system/windows/users.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      r["gid_signed"] = BIGINT(pUserIterationBuffer->usri3_primary_group_id);
+
+      r["shell"] = "";
+
+      NET_API_STATUS nStatus;
+      DWORD dwDetailedUserLevel = 23;
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Can you include a note about what this `23` is?

File: osquery/tables/system/windows/users.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                               (LPBYTE*)&pSidUserBuffer);
+      if (nStatus != NERR_Success) {
+        if (pSidUserBuffer != nullptr) {
+          NetApiBufferFree(pSidUserBuffer);
+        }
+        continue;
+      }
+
+      LPTSTR sStringSid = nullptr;
+      if (ConvertSidToStringSid(pSidUserBuffer->usri23_user_sid, &sStringSid) ==
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Mind pulling the action-expression out of the conditional?

Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: What about `pSidUserBuffer`?

File: osquery/tables/system/windows/users.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      if (wmiResults.size() != 0) {
+        std::string ret;
+        wmiResults[0].GetString("LocalPath", ret);
+        r["directory"] = ret;
+      }
+
+      LocalFree(sStringSid);
+
+      if (pUserBuffer != nullptr) {
+        NetApiBufferFree(pSidUserBuffer);
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: These are two different variables, maybe I don't understand what's happening here?

File: osquery/tables/system/windows/wmi.h [hunk 6]
```diff
@@ -14,18 +14,48 @@
...
+static std::wstring_convert<
+    std::codecvt_utf8_utf16<wchar_t, 0x10ffff, std::little_endian>>
+    converter;
+
+/**
+* @brief Windows helper function for converting narrow strings to wide
+*
+* @returns A wide string, constructed from a narrow string
+*/
+std::wstring string_to_wstring(const std::string& src);
```
Comment (theopolis) [inline osquery/tables/system/windows/wmi.h]: While you're modifying these, can you make them not snake-case?

File: osquery/tables/system/windows/users.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+#include <osquery/tables.h>
+
+#include "osquery/tables/system/windows/wmi.h"
+
+namespace osquery {
+namespace tables {
+
+void genUsers(QueryData& results) {
+  NET_API_STATUS nEnumStatus;
+  LPUSER_INFO_3 pUserBuffer = nullptr;
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: A few of these can have their scope reduced. Maybe the content within the `do`, `while` can be placed into another method that returns a `Status`?

**Comment by osqueryer** [comment]
:no_good: The commit 5549762885dcddec5c2c254d106de4b07304fd57 (Job results: [318](https://jenkins.osquery.io/job/osqueryPullRequestAudit/318/)) failed the code audit.

**Comment by facebook-github-bot-4** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/aba01f81e5044a529e7b4a59dc47f4d1c714a1a5..5549762885dcddec5c2c254d106de4b07304fd57)

**Comment by theopolis** [comment]
Looks like you need to rebase to pass the code audit.

File: osquery/tables/system/windows/users.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    WmiRequest wmiRequest(query);
+    std::vector<WmiResultItem>& wmiResults = wmiRequest.results();
+    if (wmiResults.size() != 0) {
+      std::string ret;
+      wmiResults[0].GetString("LocalPath", ret);
+      r["directory"] = ret;
+    }
+    LocalFree(sStringSid);
+
+    if (pUserBuffer != nullptr) {
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: When would pUserBuffer be null?

File: osquery/tables/system/windows/users.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      NetApiBufferFree(pSidUserBuffer);
+      pUserBuffer = nullptr;
+    }
+
+    results.push_back(r);
+    pUserIterationBuffer++;
+  }
+  if (pUserBuffer != nullptr) {
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Why is this check included?

File: osquery/tables/system/windows/users.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    pUserBuffer = nullptr;
+    pUserIterationBuffer = nullptr;
+  }
+
+  if (nEnumStatus == ERROR_MORE_DATA) {
+    LOG(INFO)
+        << "WARNING: NetUserEnum contains mode data. Results may be partial!";
+  }
+
+  if (pUserBuffer != nullptr) {
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Same

File: osquery/tables/system/windows/users.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    NetApiBufferFree(pUserBuffer);
+  }
+  return;
+}
+
+QueryData genUsers(QueryContext& context) {
+  QueryData results;
+
+  genUsers(results);
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: Why call into another function?

**Comment by facebook-github-bot-8** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/5549762885dcddec5c2c254d106de4b07304fd57..fb8edf640be452aaae2ece870ed3219b187c0c0c)

**Comment by osqueryer** [comment]
:no_good: The commit fb8edf640be452aaae2ece870ed3219b187c0c0c (Job results: [319](https://jenkins.osquery.io/job/osqueryPullRequestAudit/319/)) failed the code audit.

**Comment by facebook-github-bot-8** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/fb8edf640be452aaae2ece870ed3219b187c0c0c..a6f4f451fba7d4de8c4ebedde2cbeb8ed0d322d3)

**Comment by facebook-github-bot-1** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/a6f4f451fba7d4de8c4ebedde2cbeb8ed0d322d3..8cf85054cf4d1e5a4df107e7cfc0d16634b7e37c)

**Comment by facebook-github-bot-8** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/8cf85054cf4d1e5a4df107e7cfc0d16634b7e37c..e0c1fc51cb681ae370ac58bf18c318a0667c7b51)

File: osquery/tables/system/windows/users.cpp [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      continue;
+    }
+    r["uuid"] = sStringSid;
+    std::string query = "SELECT LocalPath FROM Win32_UserProfile where SID=\"" +
+                        std::string(sStringSid) + "\"";
+    WmiRequest wmiRequest(query);
+    std::vector<WmiResultItem>& wmiResults = wmiRequest.results();
+    if (wmiResults.size() != 0) {
+      std::string ret;
+      wmiResults[0].GetString("LocalPath", ret);
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]:
```cpp
wmiResults[0].GetString("LocalPath", r["directory"]);
```

File: osquery/tables/system/windows/users.cpp [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+    results.push_back(r);
+    pUserIterationBuffer++;
+  }
+
+  NetApiBufferFree(pUserBuffer);
+
+  if (nEnumStatus == ERROR_MORE_DATA) {
+    LOG(INFO)
+        << "WARNING: NetUserEnum contains mode data. Results may be partial!";
```
Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]:
```cpp
LOG(WARNING) << "NetUserEnum contains more data: users table may be incomplete";
```

With logging there are:
```
LOG(INFO)
LOG(WARNING)
LOG(ERROR)
```

We tend not to punctuate or use multiple statements.

Comment (theopolis) [inline osquery/tables/system/windows/users.cpp]: If you want to log something that only appears with `--verbose` then use:
```
VLOG(1)
```

If you want to log something in a debug build that is not compiled out for a release:
```
DLOG(INFO)
```

**Comment by facebook-github-bot-5** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/e0c1fc51cb681ae370ac58bf18c318a0667c7b51..39e16b6a101b19e2ecd9f7c426b347422a5c8d27)

**Comment by facebook-github-bot-1** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/39e16b6a101b19e2ecd9f7c426b347422a5c8d27..34fad6dcf3bdb813edac6de93cafca6c11e9ab8a)

**Comment by theopolis** [comment]
Why update the log to verbose only? That means there's very little chance anyone will see it.

**Comment by theopolis** [comment]
I was recommending:
```cpp
LOG(WARNING) << "NetUserEnum contains more data: users table may be incomplete";
```
:dancer:

**Comment by facebook-github-bot-0** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2506/files/34fad6dcf3bdb813edac6de93cafca6c11e9ab8a..64eb7634b498acda88415d56fc62cda75fbbee36)

**Pull request merged**